text/3550-new-range.md

```diff
@@ -0,0 +1,441 @@
+- `ops::range::IterRangeInclusive` will not implement `ExactSizeIterator` for `u16` or `i16`
+
+Those `ExactSizeIterator` impls on the legacy range types are [known to be incorrect](https://github.com/rust-lang/rust/blob/495203bf61efabecc2c460be38e1eb0f9952601b/library/core/src/iter/range.rs#L903-L936).
+
+These iterator types should each feature an associated function for getting the remaining range back:
+
+```rust
+impl<Idx> IterRange<Idx> {
+    pub fn remainder(self) -> Range<Idx>;
+}
+impl<Idx> IterRangeFrom<Idx> {
+    pub fn remainder(self) -> RangeFrom<Idx>;
+}
+impl<Idx> IterRangeInclusive<Idx> {
+    // `None` if the iterator was exhausted
+    pub fn remainder(self) -> Option<RangeInclusive<Idx>>;
+}
+```
+
+## Changed structure and API
+
+`ops::range::Range` and `ops::range::RangeFrom` will have identical structure to the existing types, with public fields for the bounds. However, `ops::range::RangeInclusive` will be changed:
+- `start` and `end` will be changed to public fields
+- `exhausted` field will be removed entirely
+
+This makes the new `RangeInclusive` the same size as `Range`.
+
+All three new types will have the same trait implementations as the legacy types, with the following exceptions:
+- NOT implement `Iterator`
+- implement `IntoIterator` directly (when `Idx: Step`)
+- implement `Copy` (when `Idx: Copy`)
+
+The following conversions between the new and legacy types will be implemented:
+```rust
+impl<Idx> From<range::Range<Idx>> for range::legacy::Range<Idx>
+impl<Idx> From<range::RangeFrom<Idx>> for range::legacy::RangeFrom<Idx>
+impl<Idx> From<range::RangeInclusive<Idx>> for range::legacy::RangeInclusive<Idx>
+
+impl<Idx> From<range::legacy::Range<Idx>> for range::Range<Idx>
+impl<Idx> From<range::legacy::RangeFrom<Idx>> for range::RangeFrom<Idx>
+// Fallible because legacy RangeInclusive can be exhausted
+impl<Idx> TryFrom<range::legacy::RangeInclusive<Idx>> for range::RangeInclusive<Idx>
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This change has the potential to cause a significant amount of churn in the ecosystem. The largest source of this churn will be cases where ranges are assumed to be `Iterator`. Unlike normal syntax churn, changes will be required to support the new range types, even on older editions.
+
+For example: while experimenting with implementing the library part of these changes, we had to modify [`quote`](https://github.com/pitaj/quote/commit/44feebf0594b255a511ff20890a7acbf4d6aeed1) and [`rustc-rayon`](https://github.com/pitaj/rustc-rayon/commit/e76e554512ce25abb48f4118576ede5d7a457918).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Just implement `Copy` on the types as-is
+
+`Copy` iterators are a large footgun. It was decided to [remove `Copy` from all iterators back in 2015](https://github.com/rust-lang/rust/pull/21809), and that decision is unlikely to be reversed.
```

Comment:
I disagree; there is a possibility that this could be reversed, but it requires new features, like #3518, to mitigate the issues it causes.

While I still think that this RFC would be nice for its own reasons, I don't think that this assumption is correct.